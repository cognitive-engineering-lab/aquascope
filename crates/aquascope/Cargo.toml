[package]
name = "aquascope"
version = "0.1.0"
license = "MIT"
edition = "2021"
authors = [
  "Gavin Gray <gavinleroy6@gmail.com>",
  "Will Crichton <wcrichto@cs.stanford.edu>"
]
description = "Permissions analyzer for Rust code"
repository = "https://github.com/cognitive-engineering-lab/aquascope"
publish = false

[package.metadata.rust-analyzer]
rustc_private = true

[features]
testing = []

[dependencies]
anyhow = "1.0.0"
log = "0.4"
itertools = "0.10.5"
serde = { version = "1.0", features = ["derive"] }
ts-rs = "6.2"
regex = "1"
fluid-let = "1.0"
rustc_utils = {version = "0.7.0-nightly-2023-08-25", features = ["graphviz", "ts-rs", "serde", "test"]}

# interpret module
miri = {git = "https://github.com/rust-lang/miri.git", rev = "63c5542edf907dd797db82c4c2979e3c4df71a8b"}
aquascope_workspace_utils = { version = "0.2", path = "../aquascope_workspace_utils" }

# testing utils
lazy_static = { version = "1.4" }

[dev-dependencies]
insta = { version = "1.22.0", features = ["json", "yaml", "redactions"] }
aquascope = { path = ".", features = ["testing"] }
env_logger = "0.9.3"
test-log = "0.2.11"
