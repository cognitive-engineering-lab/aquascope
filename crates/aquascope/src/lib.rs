#![feature(
  rustc_private,
  box_patterns,
  associated_type_defaults,
  min_specialization,
  type_alias_impl_trait,
  trait_alias,
  let_chains,
  unboxed_closures,
  exact_size_is_empty,
  type_changing_struct_update,
  extract_if,
  hash_extract_if
)]
// NOTE: these come from the clippy::pedantic group. Eventually, we'd like to deny
// the entire group (or most of it), but these are cherry picked for the time being.
#![deny(
  clippy::all,
  clippy::bool_to_int_with_if,
  clippy::case_sensitive_file_extension_comparisons,
  clippy::cloned_instead_of_copied,
  clippy::default_trait_access,
  clippy::empty_enum,
  clippy::enum_glob_use,
  clippy::expl_impl_clone_on_copy,
  clippy::explicit_deref_methods,
  clippy::filter_map_next,
  clippy::flat_map_option,
  clippy::float_cmp,
  clippy::fn_params_excessive_bools,
  clippy::from_iter_instead_of_collect,
  clippy::if_not_else,
  clippy::implicit_clone,
  clippy::inconsistent_struct_constructor,
  clippy::large_stack_arrays,
  clippy::large_types_passed_by_value,
  clippy::macro_use_imports,
  clippy::manual_assert,
  clippy::manual_let_else,
  clippy::manual_ok_or,
  clippy::manual_string_new,
  clippy::many_single_char_names,
  clippy::map_unwrap_or,
  clippy::match_bool,
  clippy::match_on_vec_items,
  clippy::match_same_arms,
  clippy::mut_mut,
  clippy::needless_for_each,
  clippy::option_option,
  clippy::similar_names
)]
// Only used for testing purposes, can we dissallow
// uncommon codepoints when not testing?
#![allow(uncommon_codepoints)]

extern crate datafrog;
extern crate either;
extern crate polonius_engine;
extern crate rustc_abi;
extern crate rustc_apfloat;
extern crate rustc_borrowck;
extern crate rustc_const_eval;
extern crate rustc_data_structures;
extern crate rustc_driver;
extern crate rustc_error_messages;
extern crate rustc_errors;
extern crate rustc_hir;
extern crate rustc_hir_pretty;
extern crate rustc_index;
extern crate rustc_infer;
extern crate rustc_interface;
extern crate rustc_macros;
extern crate rustc_middle;
extern crate rustc_mir_dataflow;
extern crate rustc_mir_transform;
extern crate rustc_serialize;
extern crate rustc_session;
extern crate rustc_span;
extern crate rustc_target;
extern crate rustc_trait_selection;
extern crate rustc_type_ir;
extern crate smallvec;

pub mod analysis;
pub mod errors;
#[allow(clippy::similar_names)]
pub mod interpreter;
#[cfg(feature = "testing")]
pub mod test_utils;
