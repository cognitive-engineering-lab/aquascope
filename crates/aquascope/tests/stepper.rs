use aquascope::test_utils;

#[test_log::test]
fn stepper() {
  test_utils::run_in_dir("steps", |path| {
    let filename = path.file_name().unwrap().to_string_lossy();
    test_utils::test_steps_in_file(path, |tag, state| {
      let f = filename.clone();
      let name = format!("{tag}@{f}");

      // Sort the nested arrays by Place string
      let mut state = state
        .into_iter()
        .map(|(s, mut vec)| {
          vec.sort_unstable_by_key(|v| v.0.clone());
          (s, vec)
        })
        .collect::<Vec<_>>();

      // Sort the outer array by line number
      state.sort_unstable_by_key(|v| v.0);

      insta::with_settings!({
        description => &name,
        omit_expression => true,
      }, {
        insta::assert_yaml_snapshot!(name, state);
      })
    });
  });
}
