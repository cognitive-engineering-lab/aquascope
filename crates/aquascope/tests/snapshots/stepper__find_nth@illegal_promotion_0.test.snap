---
source: crates/aquascope/tests/stepper.rs
description: find_nth@illegal_promotion_0.test
---
- - 0
  - - - "(*elems)[_]"
      - is_live:
          type: High
          value: true
        type_droppable:
          type: None
          value: false
        type_writeable:
          type: None
          value: false
        path_moved:
          type: None
        path_uninitialized:
          type: None
          value: false
        loan_read_refined:
          type: None
        loan_write_refined:
          type: None
        loan_drop_refined:
          type: None
        permissions:
          read:
            type: High
            value: true
          write:
            type: None
            value: false
          drop:
            type: None
            value: false
    - - "*elems"
      - is_live:
          type: High
          value: true
        type_droppable:
          type: None
          value: false
        type_writeable:
          type: None
          value: false
        path_moved:
          type: None
        path_uninitialized:
          type: None
          value: false
        loan_read_refined:
          type: None
        loan_write_refined:
          type: None
        loan_drop_refined:
          type: None
        permissions:
          read:
            type: High
            value: true
          write:
            type: None
            value: false
          drop:
            type: None
            value: false
    - - elems
      - is_live:
          type: High
          value: true
        type_droppable:
          type: High
          value: true
        type_writeable:
          type: None
          value: false
        path_moved:
          type: None
        path_uninitialized:
          type: None
          value: false
        loan_read_refined:
          type: None
        loan_write_refined:
          type: None
        loan_drop_refined:
          type: None
        permissions:
          read:
            type: High
            value: true
          write:
            type: None
            value: false
          drop:
            type: High
            value: true
    - - n
      - is_live:
          type: High
          value: true
        type_droppable:
          type: High
          value: true
        type_writeable:
          type: None
          value: false
        path_moved:
          type: None
        path_uninitialized:
          type: None
          value: false
        loan_read_refined:
          type: None
        loan_write_refined:
          type: None
        loan_drop_refined:
          type: None
        permissions:
          read:
            type: High
            value: true
          write:
            type: None
            value: false
          drop:
            type: High
            value: true

